src/Rules.Framework/Core/DictionarySlim.cs

```diff
@@ -0,0 +1,382 @@
+namespace Rules.Framework.Core
+{
+    using System;
+    using System.Collections;
+    using System.Collections.Generic;
+    using System.Runtime.CompilerServices;
+
+    internal sealed class DictionarySlim<TKey, TValue> : IDictionary<TKey, TValue>, IEnumerable, IEnumerable<KeyValuePair<TKey, TValue>>
+    {
+        private readonly IEqualityComparer<TKey> keysComparer;
+        private readonly IEqualityComparer<TValue> valuesComparer;
+        private KeyValuePair<TKey, TValue>[] entries;
+        private int lastUsedIndex;
+
+        public DictionarySlim(int size)
+        {
+            if (size <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(size), size, $"The value specified for '{nameof(size)}' must be greater than zero.");
+            }
+
+            if (typeof(TKey) == typeof(string))
+            {
+                this.keysComparer = (IEqualityComparer<TKey>)StringComparer.Ordinal;
+            }
+            else
+            {
+                this.keysComparer = EqualityComparer<TKey>.Default;
+            }
+            this.valuesComparer = EqualityComparer<TValue>.Default;
+            this.lastUsedIndex = -1;
+            this.entries = new KeyValuePair<TKey, TValue>[size];
+        }
+
+        public DictionarySlim(IDictionary<TKey, TValue> source)
+            : this(source?.Count ?? throw new ArgumentNullException(nameof(source)))
+        {
+            this.AddRange(source);
+        }
+
+        public int Count => this.lastUsedIndex + 1;
+
+        public bool IsReadOnly => false;
+
+        public ICollection<TKey> Keys => GetProjectedArrayPortion(this.entries, 0, this.Count, e => e.Key);
+
+        public ICollection<TValue> Values => GetProjectedArrayPortion(this.entries, 0, this.Count, e => e.Value)!;
+
+        public TValue this[TKey key]
+        {
+            get
+            {
+                if (key == null)
```

Comment:
You may use `key is null`

Reply:
Yes, but because `DictionarySlim<TKey, TValue>` is only used internally. `TKey` will always be a string. If it were exposed as a public type, it would be better to maintain the '==', because of operator overloading.